src/main/java/com/microsoft/aad/adal4j/SSLProxyTunnelSocketFactory.java

```diff
@@ -0,0 +1,192 @@
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package com.microsoft.aad.adal4j;
+
+import javax.net.ssl.SSLSocket;
+import javax.net.ssl.SSLSocketFactory;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.PrintStream;
+import java.net.InetAddress;
+import java.net.Socket;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Scanner;
+
+/**
+ * Supports proxy tunneling with authentication.
+ */
+public class SSLProxyTunnelSocketFactory extends SSLSocketFactory {
+    private SSLSocketFactory defaultFactory;
+
+    private String tunnelHost;
+
+    private int tunnelPort;
+
+    private ChallengeHandler challengeHandler;
+
+    /**
+     * Creates an SSLProxyTunnelSocketFactory for proxy tunneling.
+     *
+     * @param proxyHost the host of the proxy
+     * @param proxyPort the port of the proxy
+     * @param challengeHandler the handler to handle authentication challenges
+     */
+    public SSLProxyTunnelSocketFactory(String proxyHost, int proxyPort, ChallengeHandler challengeHandler) {
+        tunnelHost = proxyHost;
+        tunnelPort = proxyPort;
+        defaultFactory = (SSLSocketFactory) SSLSocketFactory.getDefault();
+
+        this.challengeHandler = challengeHandler;
+    }
+
+    public Socket createSocket(String host, int port) throws IOException {
+        return createSocket(null, host, port, true);
+    }
+
+    public Socket createSocket(String host, int port, InetAddress clientHost,
+                               int clientPort) throws IOException {
+        return createSocket(null, host, port, true);
+    }
+
+    public Socket createSocket(InetAddress host, int port) throws IOException {
+        return createSocket(null, host.getHostName(), port, true);
+    }
+
+    public Socket createSocket(InetAddress address, int port,
+                               InetAddress clientAddress, int clientPort) throws IOException {
+        return createSocket(null, address.getHostName(), port, true);
+    }
+
+    public Socket createSocket(Socket s, String host, int port,
+                               boolean autoClose) throws IOException {
+        Socket tunnel = new Socket(tunnelHost, tunnelPort);
+
+        PrintStream out = new PrintStream(tunnel.getOutputStream());
```

Comment:
PrintStream is never cleaned up after it's used. Consider wrapping it in a try-with-resources

Reply:
Will do. Thanks!

Reply:
Actually I cannot close PrintStream and InputStream - Their `close()` method will close the socket - but the socket must be kept open for the tunnel connection to remain open.